### PR TITLE
021 list overview

### PR DESCRIPTION
**Pull request by felix-horn**
Please review.

Newest updates:
- overview page with multiple shopping lists for different stores
- introducing library "immer" for manipulating database-object
- deleting entire shops

Thank you!

### CODE REVIEW

File: public/manifest.json [hunk 1]
```diff
@@ -1,25 +1,25 @@
...
-      "sizes": "512x512"
+      "sizes": "672x672"
     }
   ],
   "start_url": ".",
   "display": "standalone",
   "theme_color": "#000000",
   "background_color": "#ffffff"
-}
+}
```
Comment (JanaHaeusler) [inline public/manifest.json:25]: Reminds me to update this for my app as well :)

File: src/01.OverviewPage/01.UI-Elements/ActionButton.js [hunk 2]
```diff
@@ -0,0 +1,42 @@
+import { NavLink } from 'react-router-dom'
+import PropTypes from 'prop-types'
+import { v4 as uuid } from 'uuid'
+import styled from 'styled-components/macro'
+import PostAddIcon from '@material-ui/icons/PostAdd'
```
Comment (JanaHaeusler) [inline src/01.OverviewPage/01.UI-Elements/ActionButton.js:5]: What is a ``"Post"AddIcon`` in comparision to an ``addIcon`` ? Or is it just the predefined name?

File: src/01.OverviewPage/01.UI-Elements/ActionButton.js [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+ActionButton.propTypes = {
+  addShop: PropTypes.func.isRequired,
+  className: PropTypes.string,
+}
+
+export default function ActionButton({ addShop, className }) {
+  const newShopId = uuid()
+  return (
+    <FabStyled
```
Comment (JanaHaeusler) [inline src/01.OverviewPage/01.UI-Elements/ActionButton.js]: On first sight it is not that obvious to me what ``FabStyled`` refers to. But maybe i just do not know this abbreviation :)

File: src/01.OverviewPage/01.UI-Elements/ActionButton.md [hunk 4]
```diff
@@ -0,0 +1,3 @@
+```js padded
```
Comment (JanaHaeusler) [inline src/01.OverviewPage/01.UI-Elements/ActionButton.md:1]: You might not need to specify ``padded`` here as it is just one button.

File: src/01.OverviewPage/01.UI-Elements/ActionButton.test.js [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+      <Router>
+        <ActionButton addShop={addShopMock} />
+      </Router>
+    )
+    const button = getByTestId('action-button')
+    user.click(button)
+    expect(addShopMock).toHaveBeenCalledTimes(1)
+  })
+
+  it('links to correct href', () => {
```
Comment (JanaHaeusler) [inline src/01.OverviewPage/01.UI-Elements/ActionButton.test.js:29]: Cool idea of testing the button!

File: src/01.OverviewPage/01.UI-Elements/ActionButton.test.js [hunk 6]
```diff
@@ -0,0 +1,40 @@
+import { render } from '@testing-library/react'
+import user from '@testing-library/user-event'
+import { BrowserRouter as Router } from 'react-router-dom'
+import ActionButton from './ActionButton'
+
+describe('ActionButton', () => {
+  it('renders correctly', () => {
+    const addShopMock = jest.fn()
```
Comment (JanaHaeusler) [inline src/01.OverviewPage/01.UI-Elements/ActionButton.test.js:8]: This time you do not want to specifiy ``testProp`` at the very beginning for all tests? :)

File: src/01.OverviewPage/01.UI-Elements/CheckedItemsSummary.js [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+    <CheckedItemsSummaryStyled>
+      {lengthCheckedIds === 1
+        ? 'und 1 abgehakter Eintrag'
+        : `und ${lengthCheckedIds} abgehakte Einträge`}
+    </CheckedItemsSummaryStyled>
+  )
+}
+
+const CheckedItemsSummaryStyled = styled.p`
+  margin: 4px 2px 0 !important;
```
Comment (JanaHaeusler) [inline src/01.OverviewPage/01.UI-Elements/CheckedItemsSummary.js:19]: Ohhhh 👀

File: src/01.OverviewPage/01.UI-Elements/ListItem.js [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+  )
+}
+
+const ListItemStyled = styled.label`
+  display: flex;
+  align-items: center;
+  gap: 5px;
+`
+
+const CheckboxDiv = styled.div`
```
Comment (JanaHaeusler) [inline src/01.OverviewPage/01.UI-Elements/ListItem.js]: Maybe just ``Checkbox``or ``CheckboxContainer`` ,``CheckboxWrapper`` or something else? I think you have not used element descriptions (div etc.) in component names elsewhere, have you?

File: src/01.OverviewPage/01.UI-Elements/ListItem.md [hunk 9]
```diff
@@ -0,0 +1,3 @@
+```js padded
```
Comment (JanaHaeusler) [inline src/01.OverviewPage/01.UI-Elements/ListItem.md:1]: Same here :)

File: src/01.OverviewPage/02.Components/ShopCard.test.js [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+      },
+    },
+  },
+}
+
+describe('ShopCard', () => {
+  it('renders correctly', () => {
+    const { container } = render(
+      <Router>
+        <ShopCard shopId="x" database={testDatabase} />
```
Comment (JanaHaeusler) [inline src/01.OverviewPage/02.Components/ShopCard.test.js:48]: Different to line 57 -> ``shopId = "x"``  ``shopId={'x'}``

File: src/01.OverviewPage/OverviewPage.test.js [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+    const button = getByTestId('action-button')
+    user.click(button)
+    expect(addShopMock).toHaveBeenCalledTimes(1)
+  })
+
+  it('links to correct href', () => {
+    const addShopMock = jest.fn()
+    const { getByTestId } = render(
+      <Router>
+        <OverviewPage database={testDatabase} addShop={addShopMock} />
```
Comment (JanaHaeusler) [inline src/01.OverviewPage/OverviewPage.test.js:71]: You could think of a general ``testProps`` here as well.

File: src/02.ShopPage/01.UI-Elements/AddItemButton.js [hunk 12]
```diff
@@ -4,20 +4,25 @@ import AddIcon from '@material-ui/icons/Add'
...
       <AddIconStyled />
       <ButtonTextStyled>Listeneintrag</ButtonTextStyled>
     </AddItemButtonStyled>
   )
 }
 
 const AddItemButtonStyled = styled.button`
-  margin-left: 27px;
-  margin-bottom: 20px;
+  //margin-left: 27px;
```
Comment (JanaHaeusler) [inline src/02.ShopPage/01.UI-Elements/AddItemButton.js]: Do you need them? :)

File: src/02.ShopPage/01.UI-Elements/ListItem.js [hunk 13]
```diff
@@ -52,7 +52,7 @@ export default function ListItem({
   )
 
   function handleCheckboxToggle() {
-    clearTimeout(raceConditionTimer)
+    //clearTimeout(raceConditionTimer)
```
Comment (JanaHaeusler) [inline src/02.ShopPage/01.UI-Elements/ListItem.js:55]: Same here :)

File: src/02.ShopPage/01.UI-Elements/ListItem.js [hunk 14]
```diff
@@ -70,12 +70,12 @@ export default function ListItem({
   }
 
   function hideDeleteIcon() {
-    raceConditionTimer = setTimeout(() => setIsDeleteIconShown(false), 150)
-    return () => clearTimeout(raceConditionTimer)
+    raceConditionTimer = setTimeout(() => setIsDeleteIconShown(false), 0)
+    //return () => clearTimeout(raceConditionTimer)
```
Comment (JanaHaeusler) [inline src/02.ShopPage/01.UI-Elements/ListItem.js:74]: Is it a comment or code?

File: src/02.ShopPage/01.UI-Elements/ShopTitle.test.js [hunk 15]
```diff
@@ -0,0 +1,40 @@
...
+
+const testProps = {
+  shopId: 'x',
+  database: testDatabase,
+  changeTitle: jest.fn(),
+}
+
+describe('ShopTitle', () => {
+  it('shows the correct title', () => {
+    const props = { ...testProps, title: 'shopX' }
```
Comment (JanaHaeusler) [inline src/02.ShopPage/01.UI-Elements/ShopTitle.test.js]: ``title`` is not included in ``testProps``

File: src/02.ShopPage/02.Components/Menu.js [hunk 16]
```diff
@@ -0,0 +1,42 @@
...
+  return (
+    <MenuStyled className={className}>
+      <DeleteEntry
+        onClick={deleteShop}
+        exact
+        to="/"
+        data-testid="delete-button"
+      >
+        <DeleteForeverIcon />
+        Liste unwideruflich löschen
```
Comment (JanaHaeusler) [inline src/02.ShopPage/02.Components/Menu.js]: un­wi­der­ruf­lich

File: src/02.ShopPage/02.Components/Menu.test.js [hunk 17]
```diff
@@ -0,0 +1,40 @@
...
+        <Menu deleteShop={deleteShopMock} />
+      </Router>
+    )
+    const deleteButton = getByTestId('delete-button')
+    user.click(deleteButton)
+    expect(deleteShopMock).toHaveBeenCalledTimes(1)
+  })
+
+  it('links to correct href', () => {
+    const deleteShopMock = jest.fn()
```
Comment (JanaHaeusler) [inline src/02.ShopPage/02.Components/Menu.test.js:30]: Here same thought again: Do you want to stick with general ``testProps`` or define it within each test?

File: src/02.ShopPage/02.Components/UncheckedList.js [hunk 18]
```diff
@@ -15,63 +16,57 @@ UncheckedList.propTypes = {
...
-                )
-              })}
-              {provided.placeholder}
-            </UncheckedListStyled>
-          )}
-        </Droppable>
-      </DragDropContext>
-    </>
+    <DragDropContext
+      onDragStart={giveHapticFeedback}
```
Comment (JanaHaeusler) [inline src/02.ShopPage/02.Components/UncheckedList.js:33]: Awesome feature :)

File: src/02.ShopPage/02.Components/UncheckedList.test.js [hunk 19]
```diff
@@ -49,18 +51,27 @@ const testProps = {
...
-    const { container } = render(
-      <UncheckedList {...testProps} database={testDatabase} />
-    )
+    const { container } = render(<UncheckedList {...testProps} />)
     expect(container.firstChild).toMatchSnapshot()
   })
+
+  it('only renders listItems of its shop', () => {
+    const { queryByDisplayValue, getByDisplayValue } = render(
+      <UncheckedList {...testProps} shopId={'x'} database={testDatabase} />
```
Comment (JanaHaeusler) [inline src/02.ShopPage/02.Components/UncheckedList.test.js]: ``shopId`` and ``database`` are already in your ``testProps``.

File: src/02.ShopPage/02.Components/UncheckedList.test.js [hunk 20]
```diff
@@ -49,18 +51,27 @@ const testProps = {
...
+    )
+    const itemA = getByDisplayValue(/itemA/i)
+    const itemC = queryByDisplayValue(/itemC/i)
+    expect(itemA).toBeInTheDocument()
+    expect(itemC).not.toBeInTheDocument()
+  })
+
   it('only renders unchecked listItems', () => {
     const { queryByDisplayValue, getByDisplayValue } = render(
       <UncheckedList {...testProps} database={testDatabase} />
```
Comment (JanaHaeusler) [inline src/02.ShopPage/02.Components/UncheckedList.test.js:70]: Here as well.

File: src/hooks/useDatabase.js [hunk 21]
```diff
@@ -28,122 +34,137 @@ export default function useDatabase() {
...
-      },
-    })
+    setDatabase(
+      produce(database, (draft) => {
+        draft.shops.byId[shopId].items = idsShopItems
+      })
+    )
+  }
+
+  /* ---------------------- delete ListItem and undo delete --------------------------- */
```
Comment (JanaHaeusler) [inline src/hooks/useDatabase.js:106]: Not sure if this should probably deleted...

File: src/styles/GlobalStyle.js [hunk 22]
```diff
@@ -4,12 +4,13 @@ export default createGlobalStyle`
 
 :root {
     
-
+    --warning-red: #de5246;
     --almost-black: #212121; //font-color
```
Comment (JanaHaeusler) [inline src/styles/GlobalStyle.js:8]: Would probably change the naming, in order to know, what the color is for and not to have to write comments. For example: ``--font-color: #212121``

**Review by JanaHaeusler — COMMENTED** [review]
A lot of hard coding work and a very complex and interesting code in a good way. :) Left some thoughts. Hopefully they are of any help. You could maybe think about whether to go with general ``testProps`` in your tests (when there are some of them) or not. 👏🏻 :)

File: src/01.OverviewPage/01.UI-Elements/ActionButton.js (hunk 2 above)
Comment (felix-horn) [reply src/01.OverviewPage/01.UI-Elements/ActionButton.js:5]: Exactly. That's what Google calls it :)

File: src/01.OverviewPage/01.UI-Elements/ActionButton.js (hunk 3 above)
Comment (felix-horn) [reply src/01.OverviewPage/01.UI-Elements/ActionButton.js]: Ah, you are right. It means "Floating Action Button". I used to import it from the material-ui as well now I don't and so I could give it a different name. Thanks.

File: src/01.OverviewPage/01.UI-Elements/ActionButton.md (hunk 4 above)
Comment (felix-horn) [reply src/01.OverviewPage/01.UI-Elements/ActionButton.md:1]: What does it mean anyways...? :)

**Pull request merged**